rapids_build_backend/impls.py

```diff
@@ -210,6 +210,36 @@ def _edit_pyproject(config):
         shutil.move(bkp_pyproject_file, pyproject_file)
 
 
+def _check_setup_py(setup_py_contents: str) -> None:
+    """
+    ``setuptools.get_requires_for_build_wheel()`` executes setup.py if it exists,
+    to check for dependencies in ``setup_requires`` (passed to ``setuptools.setup()``).
+
+    That's a problem for rapids-build-backend, as at the point where that's invoked,
+    its recalculated list of build dependencies (modified in ``_edit_pyproject()``)
+    haven't yet been installed.
+
+    If any of them are imported in ``setup.py``, those imports will fail.
+
+    This function raises an exception if it detects ``setup_requires`` being used in
+    a ``setup.py``, to clarify that ``rapids-build-backend`` can't support that case.
+
+    ref: https://github.com/rapidsai/rapids-build-backend/issues/39
+    """
+
+    # pattern = "any use of 'setup_requires' on a line that isn't a comment"
+    setup_requires_pat = r"^(?!\s*#+).*setup_requires"
```

Comment:
ugh reason 1000 that no one should use `"""` strings as comments `*shakes fist*`.

You're totally right though, I didn't consider that. I'll adjust for that.

Reply:
Or we can just not use a regex and instead parse the Python code. This ChatGPT response worked in my local testing:

```python
import ast


def find_setup_requires_in_calls(filename):
    with open(filename, "r") as file:
        tree = ast.parse(file.read(), filename=filename)

    class CallVisitor(ast.NodeVisitor):
        def __init__(self):
            self.found = False

        def visit_Call(self, node):
            for keyword in node.keywords:
                if keyword.arg == "setup_requires":
                    self.found = True
                    print(
                        f"'setup_requires' found in a call to function: {ast.dump(node.func)} at line {node.lineno}"
                    )
            self.generic_visit(node)

    visitor = CallVisitor()
    visitor.visit(tree)

    if not visitor.found:
        print("No 'setup_requires' argument found in any function call.")


# Replace 'your_file.py' with the path to the file you want to check.
find_setup_requires_in_calls("your_file.py")
```

Reply:
I haven't used `ast`, but I understand it's purpose and what it's used for.

This looks to me like it'd introduce a bigger surface area for bugs than using a regex...

* are there cases where the output of `ast.parse()` takes up a non-trivial amount of memory? 
* right now this has `for keyword in node.keywords`... does this also need to be extended for the use of positional arguments?
* will this catch the case where you batch up all the arguments in a dictionary and then pass them through like `setup(**opts)`?

If and of you are more familiar than me with `ast` and think these things I'm talking about aren't worth worrying about, then I'll happily switch to something like this.

Reply:
There are pros and cons. If we're worried about multi-line comments giving false positives, we should use a proper parser rather than hacked together regular expressions with multiple passes.

* are there cases where the output of `ast.parse()` takes up a non-trivial amount of memory?
  * Maybe, but it'd be unsafe for Python to load that `setup.py` if so. This is not a case I would worry about.
* right now this has `for keyword in node.keywords`... does this also need to be extended for the use of positional arguments?
  * I've never seen a `setup.py` use positional arguments, and if it does, we'd be unable to use the regex solution anyway. Probably not a situation we should worry about.
* will this catch the case where you batch up all the arguments in a dictionary and then pass them through like `setup(**opts)`?
  * Nope, that won't be supported by an ast solution. This would be a win for regex. But if you're worried about using `"""` for comments, it'd be really hard to identify `"Comments that are just strings in the code -- which is all that the triple-quotes strings are, anyway"`

I don't think we should try to overengineer this. Either a single-pass regex or simplest possible AST is what I would recommend, and of course document that we don't support `setup_requires`.

Reply:
> *[the memory footprint] is not a case I would worry about.*

ok, thanks.

> *if [setup.py does use positional args] we'd be unable to use the regex solution anyway.*

Right, I was thinking that this would be a situation where maybe the AST approach would provide some benefit over regex in exchange for the added complexity.

I do agree that it's not a case that's worth worrying about.

> *I don't think we should try to overengineer this*

I agree. I think we are coming to this with different opinions about what that word means though.

My perspective is that for the types of situations we expect this project to be used in, introducing syntax-tree parsing of the `setup.py` for this purpose is "overengineering".

But I'd prefer that to doing a multi-pass regex to handle the case of people using `"""` strings as comments.

I'll update this to use the AST approach.

Reply:
I agree with James, using `ast` already feels like overengineering here. One alternative is to simply ignore the multi-line comment case, see what happens, and if we ever have a project using RBB that actually breaks this we can add a config option to tell RBB to ignore this check. Lowest engineering effort for a case that we don't really care about.

Reply:
My viewpoint is basically this:
- I am willing to accept false positives to simplify the engineering effort, and we can expand or improve or optionally disable things later if needed
- I prefer the simplest possible solution (probably a single-pass minimal regex as implemented originally) unless someone thinks it is insufficient
- Vyas started this thread to discuss a flaw in this solution, which I took to mean that he does not think the simple regex is sufficient, or he has a different tolerance for false positives
- Otherwise, I prefer ast over multi-pass regex as I think it is the next-simplest solution

Reply:
I started the thread because I recognized the issue but didn't think _that_ hard about how big of an issue the problem really was. Now that we've had this discussion about the possible solutions, I'm leaning towards the effort being not worthwhile and that we should just accept the false positives. It's not very restrictive, it just means projects will fail to build if `setup_requires` is found inside a multi-line comment and they can then change the comment syntax if needed. So let's skip trying to fix this.

Reply:
Alright thanks for that, I agree.

False positives here are not that costly.. they result in a big loud error *(not, say, silently-wrong behavior)*, and they happen only at **build** time which means no RAPIDS user should ever be affected.

Reply:
> and of course document that we don't support setup_requires.

@bdice want to be sure you know I didn't miss this comment.

Proposed documentation on that here: https://github.com/rapidsai/rapids-build-backend/pull/41/commits/1c7c9e5071b514e93850fe43ca257a01fcdccd27